cluster.go

```diff
@@ -1,90 +1,142 @@
 package main
 
-import "fmt"
+import (
+	"fmt"
+	"math/rand"
+	"sync"
+)
 import "time"
-import "math/rand"
 
 const ClusterSize = 8
-const ElectionTimeOut = 1000
+const ElectionTimeOut = 2 * 1000 // in milliseconds
 
 type Vote struct {
 	Term int
 	VoteFor int
 	Responses chan bool
 }
 
-
 func initCluster(done chan bool) {
 
 	var voteChannels [ClusterSize]chan Vote
+	var leaderCommunicationChannel [ClusterSize] chan LogEntry
 
 	// Spawn 8 nodes (all followers to start)
 	for i := 0; i < ClusterSize; i++ {
 		// initialize state as followers
-		state := ServerState{i, 0, -1, []LogEntry{}}
+		state := ServerState{i, 0, -1, []LogEntry{}, FollowerRole}
 
 		voteChannels[i] = make(chan Vote)
+		leaderCommunicationChannel[i] = make(chan LogEntry)
 
-		go startServer(state, &voteChannels, done)
+		go startServer(state, &voteChannels, &leaderCommunicationChannel, done)
 	}
 }
 
-
-func startServer(state ServerState, voteChannels *[ClusterSize]chan Vote, done chan bool) {
-	//start election timer
-	electionTime := time.NewTimer(time.Duration(ElectionTimeOut) * time.Millisecond)
-
-	elect(state, voteChannels, electionTime)
-	
-	done <- true
+/* Creates a server in the cluster. Structured via https://pdos.csail.mit.edu/6.824/labs/raft-structure.txt
+ *
+ */
+func startServer(
+	state ServerState,
+	voteChannels *[ClusterSize]chan Vote,
+	leaderCommunicationChannels *[ClusterSize]chan LogEntry,
+	done chan bool,
+) {
+	isElection := false
+	electionThreadSleepTime := time.Millisecond * 50
+	timeSinceLastUpdate := time.Now() //update includes election or message from leader
+	serverStateLock := new(sync.Mutex)
+
+	/* Election Timer: Checks if timeout is surpassed and starts election. Timeout is reached when:
+	 * 1. no message from leader or
+	 * 2. when election took too long (e.g. due to tie / no leader elected)
+	 */
+	go func () {
+		for {
+			timeElapsed := time.Now().Sub(timeSinceLastUpdate)
+			if timeElapsed.Milliseconds() > ElectionTimeOut {
+				isElection = true
+				timeSinceLastUpdate = time.Now()
+				go elect(&state, voteChannels, leaderCommunicationChannels)
+			}
+			time.Sleep(electionThreadSleepTime)
+		}
+	}()
```

Comment:
Not totally clear on how you're waiting for messages from the leader while also waiting on timer

Reply:
So, from your previous comment above I took out the timer and just have a timestamp from the last update -- either a message from the leader or the start of an election. If either goes over the `ElectionTimeOut` then it's time to run another election. This covers the case where an election took too long and allowing me to take out the timer that existed before